src/finalfusion/embeddings.py

```diff
@@ -426,6 +541,20 @@ def __iter__(self):
             return zip(self._vocab.words, self._storage, self._norms)
         return zip(self._vocab.words, self._storage)
 
+    def _similarity(self, query: np.ndarray, k: int,
+                    skips: Set[str]) -> List['SimilarityResult']:
+        sims = self.storage[:len(self.vocab)].dot(query)
+        skip_indices = [
+            skip for skip in (self.vocab.word_index.get(skip)
+                              for skip in skips) if skip is not None
+        ]
+        indices = sims.argsort()[::-1]
```

Comment:
This does an *O(n log n)* sort. A good binary heap implementation should be faster, since insertions are *O(1)* in the average case. Also, the worst case would only be *O(log k)* if the heap is at most *k* items. So, inserting all items in the heap should be *O(n)* on average or *O(n log k)* worst case.

(A more niche approach that would be faster is something like quicksort, where you'd only guarantee the first *k* items to be sorted. numpy seems to have a function for this:

https://numpy.org/doc/stable/reference/generated/numpy.argpartition.html

I think you could either use argpartition with an array, or you could use it to get the top-k elements and then sort, so that the running time of the sort is only *O(k log k)*.)

Reply:
Went with `argpartition` and sorting a heap after partioning.